cpp/src/traversal/two_hop_neighbors.cu

```diff
@@ -29,154 +31,96 @@
-                                                    exsum_degree + 1,
-                                                    exsum_degree + num_edges + 1,
-                                                    exsum_degree + 1);
+                           d_exsum_degree + 1,
+                           d_exsum_degree + graph.number_of_edges + 1,
+                           d_exsum_degree + 1);
 
     // Copy out the last value to get the size of scattered output
-    IndexType output_size;
-    cudaMemcpy(&output_size, &exsum_degree[num_edges], sizeof(IndexType), cudaMemcpyDefault);
+    ET output_size = exsum_degree[graph.number_of_edges];
 
     // Allocate memory for the scattered output
-    ALLOC_TRY(&second_pair, sizeof(IndexType) * output_size, stream);
-    ALLOC_TRY(&first_pair, sizeof(IndexType) * output_size, stream);
+    rmm::device_vector<VT> first_pair(output_size);
+    rmm::device_vector<VT> second_pair(output_size);
+
+    VT *d_first_pair = first_pair.data().get();
+    VT *d_second_pair = second_pair.data().get();
 
     // Figure out number of blocks and allocate memory for block bucket offsets
-    IndexType num_blocks = (output_size + TWO_HOP_BLOCK_SIZE - 1) / TWO_HOP_BLOCK_SIZE;
-    ALLOC_TRY(&block_bucket_offsets, sizeof(IndexType) * (num_blocks + 1), stream);
+    ET num_blocks = (output_size + TWO_HOP_BLOCK_SIZE - 1) / TWO_HOP_BLOCK_SIZE;
+    rmm::device_vector<ET> block_bucket_offsets(num_blocks+1);
+
+    ET *d_block_bucket_offsets = block_bucket_offsets.data().get();
 
     // Compute the block bucket offsets
     dim3 grid, block;
     block.x = 512;
-    grid.x = min((IndexType) MAXBLOCKS, (num_blocks / 512) + 1);
-    compute_bucket_offsets_kernel<<<grid, block, 0, nullptr>>>(exsum_degree,
-                                                               block_bucket_offsets,
-                                                               num_edges,
+    grid.x = min((ET) MAXBLOCKS, (num_blocks / 512) + 1);
+    compute_bucket_offsets_kernel<<<grid, block, 0, nullptr>>>(d_exsum_degree,
+                                                               d_block_bucket_offsets,
+                                                               graph.number_of_edges,
                                                                output_size);
-    cudaMemcpy(&block_bucket_offsets[num_blocks], &num_edges, sizeof(IndexType), cudaMemcpyDefault);
+
+    block_bucket_offsets[num_blocks] = graph.number_of_edges;
 
     // Scatter the expanded edge lists into temp space
-    grid.x = min((IndexType) MAXBLOCKS, num_blocks);
-    scatter_expand_kernel<<<grid, block, 0, nullptr>>>(exsum_degree,
-                                                       indices,
-                                                       offsets,
-                                                       block_bucket_offsets,
-                                                       num_verts,
+    grid.x = min((ET) MAXBLOCKS, num_blocks);
```

Comment:
grid is a `dim3` which is afaik based on `uint3`. 
`ET` is expected to be 64 bit soon and `MAXBLOCKS` is a constant that does not require 64bits. It is probably safer to avoid involving 64 bit here when dealing with grid sizes.

Reply:
If ET becomes 64-bit at some point, then somewhere in the derivation of grid.x we need to reduce from 64-bit to 32-bit.  I had deferred that to the last moment (here).  I could make num_blocks a 32-bit integer.  I felt like deferring to this point was more future-proof - in case dim3 eventually changes to a different type we don't have to worry about changing types elsewhere.